openverse_catalog/dags/providers/provider_api_scripts/time_delineated_provider_data_ingester.py

```diff
@@ -0,0 +1,238 @@
+        pass
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # This class is used to generate timestamp intervals over a 24-hour period,
+        # so it should only be used for a dated DAG.
+        assert (
+            self.date is not None
+        ), f"{self.__class__.__name__} should only be used for dated DAGs."
+
+        # A flag that is True only when we are processing the first batch of data in
+        # a new iteration.
+        self.new_iteration = True
+        # Use to keep track of the number of records we've fetched from the API so far,
+        # specifically in this iteration.
+        self.fetched_count = 0
+
+    @staticmethod
+    def format_ts(timestamp):
+        return timestamp.isoformat().replace("+00:00", "Z")
+
+    @staticmethod
+    def _get_timestamp_query_params_list(
+        start_date: datetime, end_date: datetime, number_of_divisions: int
+    ) -> list[tuple[datetime, datetime]]:
+        """
+        Given a start_date and end_date, returns a list of timestamp pairs that
+        divides the time interval between them into equal portions, with the
+        number of portions determined by `number_of_divisions`.
+
+        Required Arguments:
+        start_date:          datetime to be considered start of the interval
+        end_date:            datetime to be considered end of the interval
+        number_of_divisions: int number of portions to divide the interval into.
+        """
+        seconds_in_time_slice = (end_date - start_date).total_seconds()
+        portion = int(seconds_in_time_slice / number_of_divisions)
+        # Double check that the division resulted in even portions
+        if seconds_in_time_slice % number_of_divisions:
+            raise ValueError(
+                f"The time slice from {start_date} to {end_date} cannot be divided "
+                f"evenly into {number_of_divisions} parts!"
+            )
+
+        # Generate the start/end timestamps for each 'slice' of the interval
+        return [
+            (
+                start_date + timedelta(seconds=i * portion),
+                start_date + timedelta(seconds=(i + 1) * portion),
+            )
+            for i in range(number_of_divisions)
+        ]
+
+    @abstractmethod
+    def get_record_count_from_response(self, response_json):
+        # Given a response_json, return the response_count. This should not be the
+        # number of records in the response, but the reported count.
+        pass
+
+    def _get_record_count(self, start: datetime, end: datetime, **kwargs) -> int:
+        """Get the number of records returned by the API for a particular interval."""
+        query_params = self.get_next_query_params(
+            prev_query_params=None, start_ts=start, end_ts=end, **kwargs
+        )
+        response_json = self.get_response_json(query_params)
+
+        return int(self.get_record_count_from_response(response_json))
```

Comment:
Would it make sense for `get_record_count_from_response` to be expected to return an `int` rather than convert it here? Especially since that's expected to be implemented by subclasses, that seems like it might be a good thing to enforce at that level (rather than deal with cases where that return value is `None` and we inadvertently raise a `NoneType` error).

Reply:
Totally agreed! I changed this at the last minute when I realized Flickr was returning a string, but we should enforce this in the subclass instead 👍